src/main.cpp

```diff
@@ -7622,70 +7622,72 @@ bool SendMessages(CNode *pto)
+                // However we will still send them block inventory in the case they are a pruned node or wallet
+                // waiting for block announcements, therefore we have to check each inv in pto->vInventoryToSend.
+                bool chokeTxInv = (pto->nActivityBytes == 0 && (nNow / 1000000 - pto->nTimeConnected) > 120);
+                LOCK(pto->cs_inventory);
 
-                // BU - here we only want to forward message inventory if our peer has actually been requesting useful
-                // data or
-                //      giving us useful data.  We give them 2 minutes to be useful but then choke off their inventory.
-                //      This
-                //      prevents fake peers from connecting and listening to our inventory while providing no value to
-                //      the network.
-                //      However we will still send them block inventory in the case they are a pruned node or wallet
-                //      waiting for block
-                //      announcements, therefore we have to check each inv in pto->vInventoryToSend.
-                if (inv.type == MSG_TX && pto->nActivityBytes == 0 && (GetTime() - pto->nTimeConnected) > 120)
-                {
-                    LogPrint("evict", "choking off tx inventory for %s time connected %d\n", pto->addr.ToString(),
-                        GetTime() - pto->nTimeConnected);
-                    continue;
-                }
+                vInvSend.reserve(pto->vInventoryToSend.size());
+                // about 1/4 of the nodes should end up in this list, so over-allocate
+                vInvWait.reserve(pto->vInventoryToSend.size() / 3);
 
-                // trickle out tx inv to protect privacy
-                if (inv.type == MSG_TX && !fSendTrickle)
+                // Make copy of vInventoryToSend while cs_inventory is locked but also ignore some tx and defer others
+                BOOST_FOREACH (const CInv &inv, pto->vInventoryToSend)
                 {
-                    // 1/4 of tx invs blast to all immediately
-                    static uint256 hashSalt;
-                    if (hashSalt.IsNull())
-                        hashSalt = GetRandHash();
-                    uint256 hashRand = ArithToUint256(UintToArith256(inv.hash) ^ UintToArith256(hashSalt));
-                    hashRand = Hash(BEGIN(hashRand), END(hashRand));
-                    bool fTrickleWait = ((UintToArith256(hashRand) & 3) != 0);
-
-                    if (fTrickleWait)
+                    if (inv.type == MSG_TX)
                     {
-                        vInvWait.push_back(inv);
-                        continue;
+                        if (chokeTxInv)
+                            continue;
+                        // skip if we already know abt this one
+                        if (pto->filterInventoryKnown.contains(inv.hash))
+                            continue;
+                        if (!fSendTrickle)
+                        {
```

Comment:
I would have kept this comment:
```
// 1/4 of tx invs blast to all immediately
```

Reply:
I added the comment back in.  I noticed that the new code was actually blasting 3/4 of the invs immediately so I also changed that to match the original code (and comment).

I also noticed that the "fSendTrickle" semantics is opposite of the meaning of the variable:
in the original code, if fSendTrickle was FALSE, it would activate the trickle functionality.  So for clarity, I switched the true/false of this boolean in all places.  This refactor does not change program semantics, it just aligns the english meaning of the variable with its value.

Reply:
ok, makes sense. going to have a look at code reorg.